### ISSUE TITLE
Extend keyboard manager to support cheat codes

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#4159]
Elsewhere, @twMat has been seeking a way to make his tiddlyspot hosted wiki present a different UI for him versus other, ordinary users (as I understand it, please correct if I'm wrong).

A somewhat traditional solution to this problem might be to add support for [cheat codes](https://en.wikipedia.org/wiki/Cheating_in_video_games#Cheat_codes): a magic key combination of multiple consecutive keypresses that triggers an action that modifies the UI. The tiddler(s) holding the state could be stored in browser storage, meaning that @twMat would only have to repeat the magic key combination once per browser.

So, @BurningTreeC this might be one for you: could the keyboard manager track enough state that it could use sequences of keyboard shortcuts to trigger an action?

**Comment by BurningTreeC** [comment]
@Jermolene , this appears to be a huge rewrite of the mechanism I'm not able to do

**Comment by twMat** [comment]
>a way to make his tiddlyspot hosted wiki present a different UI for him versus other, ordinary users

Yes. Much appreciated that you bring this up, and that's a nice proposed solution which should be of use for anyone who wants to control/trigger *anything* locally for whatever reason.

>this appears to be a huge rewrite of the mechanism I'm not able to do

@Jermolene - do you have any specific thoughts or pointers that would ease understanding for how to approach this?

**Comment by Jermolene** [comment]
> do you have any specific thoughts or pointers that would ease understanding for how to approach this?

The approach I was thinking would require a change to the core keyboard handling. Each shortcut would get an array of keys that trigger it, instead of a single one. The system would keep track of the last X keypresses, and match against that against the array instead of a single keypress.

**Comment by AnthonyMuscio** [comment]
Would an obscure three key combination not be enough. Can that be handled by the current shortcut methods?, I wanted something similar to trigger a save so I could conceal the save mechanism then save the wiki (for read only users).

**Comment by pmario** [comment]
IMO has been implemented with https://tiddlywiki.com/#Alternative%20page%20layouts
